common/test/java/com/couchbase/lite/LiveQueryTest.java

```diff
@@ -267,54 +217,49 @@ public void testChangeParameters() throws CouchbaseLiteException, InterruptedExc
         Parameters params = new Parameters();
         params.setInt("VALUE", 2);
         query.setParameters(params);
 
-        ListenerToken token = query.addChangeListener(testSerialExecutor, change -> globalLatch.countDown());
+        ListenerToken token = query.addChangeListener(
+            testSerialExecutor,
+            change -> {
+                ResultSet rs = change.getResults();
+                //  query should only be notified 2 times:
+                //  1. query first gets doc 2, the rs size is 1
+                //  2. after param changes to 1, query gets a new rs that has doc 1 and 2, rs size is now 2
+                //  query should not be notified when doc 0 is added to the db
+                if (rs != null && rs.allResults().size() == count[0] + 1) {
+                    latch[count[0]++].countDown();
+                    rs.close();
+                }
+            });
         try {
-            assertTrue(globalLatch.await(LONG_TIMEOUT_SEC, TimeUnit.SECONDS));
-
-            globalLatch = new CountDownLatch(1);
+            assertTrue(latch[0].await(TOLERABLE_DATABASE_CHANGE_DELAY, TimeUnit.MILLISECONDS));
 
             params = new Parameters();
             params.setInt("VALUE", 1);
             query.setParameters(params);
 
-            assertTrue(globalLatch.await(SLOP_MS, TimeUnit.MILLISECONDS));
-        }
-        finally {
-            query.removeChangeListener(token);
-        }
-    }
-
-    // https://github.com/couchbase/couchbase-lite-android/issues/1606
-    @Test
-    public void testRemovingLiveQuery() throws CouchbaseLiteException, InterruptedException {
-        int n = 1;
-        newQuery(n);
-        try {
-            // creates doc1 -> first query match
-            createDocNumbered(n++);
-            assertTrue(globalLatch.await(LONG_TIMEOUT_SEC, TimeUnit.SECONDS));
+            // VALUE changes to 1, query now gets a new rs for doc 1 and 2
+            assertTrue(latch[1].await(TOLERABLE_DATABASE_CHANGE_DELAY, TimeUnit.MILLISECONDS));
 
-            // create doc2 -> update query match
-            createDocNumbered(n++);
-            assertTrue(globalLatch.await(LONG_TIMEOUT_SEC, TimeUnit.SECONDS));
-
-            // create doc3 -> update query match
-            createDocNumbered(n);
-            assertTrue(globalLatch.await(LONG_TIMEOUT_SEC, TimeUnit.SECONDS));
-        }
-        finally {
-            globalQuery.removeChangeListener(globalToken);
+            // This doc does not meet the condition of the query, thus query should not get notified
+            createDocNumbered(0);
+            assertFalse(latch[2].await(TOLERABLE_DATABASE_CHANGE_DELAY * 2, TimeUnit.MILLISECONDS));
         }
+        finally { query.removeChangeListener(token); }
     }
 
     // CBL-2344: Live query may stop refreshing
-    @Ignore("Fails during re-implementation of LiveQuery")
+    @SuppressWarnings("ConstantConditions")
```

Comment:
You should probably apply this to the whole class.  As I say, never, never in production... In the tests, it is fine.